cfme/utils/ssh.py

```diff
@@ -307,20 +304,39 @@ def run_command(
             session.exec_command(command)
             stdout = session.makefile()
             stderr = session.makefile_stderr()
+
+            def write_output(line, file):
+                output.append(line)
+                if self._streaming:
+                    file.write(line)
+
             while True:
+                # While the program is running loop through collecting line by line so that we don't
+                # fill the buffers up without a newline
                 if session.recv_ready:
-                    for line in stdout:
-                        output.append(line)
-                        if self._streaming:
-                            self.f_stdout.write(line)
+                    try:
+                        line = stdout.next()
+                        write_output(line, self.f_stdout)
+                    except StopIteration:
+                        pass
 
                 if session.recv_stderr_ready:
-                    for line in stderr:
-                        output.append(line)
-                        if self._streaming:
-                            self.f_stderr.write(line)
+                    try:
+                        line = stderr.next()
+                        write_output(line, self.f_stderr)
+                    except StopIteration:
+                        pass
 
                 if session.exit_status_ready():
+                    # When the program finishes, we need to grab the rest of the output that is left
+                    # Though it's possible, we should have read enough of the buffers so that we can
+                    # just dump the rest.
+                    if session.recv_ready:
+                        for line in stdout:
```

Comment:
What happens if the last line does not send an EOL character?

Reply:
I mean if the last bit of output from the remote program does not have an EOL ending it.

Reply:
@jamesooden So stdout is a paramiko ChannelFile, which is a paramiko BufferedFile that behaves like a python file.

https://github.com/paramiko/paramiko/blob/master/paramiko/channel.py#L669

Here, in the recv method, which is used for ChannelFile.read(), the docblock includes the description of the behavior when zero length string is received.

It looks to me like if there is no EOL, it won't necessarily matter as it will just handle it as the file has been closed.

Some interesting points in paramiko: 
https://github.com/paramiko/paramiko/blob/master/paramiko/file.py#L291
https://github.com/paramiko/paramiko/blob/master/paramiko/file.py#L119